File: src/axis_parser.js
"use strict";

var SRC = files.path(engines.myEngine().cwd() + "/src");
var models = require(SRC + "/models.js");

function parseTimeToSeconds(raw) {
  var match = String(raw).trim().match(/^(\d):([0-5]\d)$/);
  if (!match) {
    throw new Error("时间格式必须为 M:SS：" + raw);
  }
  return parseInt(match[1], 10) * 60 + parseInt(match[2], 10);
}

function parseKeyValue(line, lineNumber) {
  var index = line.indexOf("=");
  if (index < 0) {
    throw new Error("第" + lineNumber + "行缺少等号：" + line);
  }
  return {
    key: line.slice(0, index).trim(),
    value: line.slice(index + 1).trim()
  };
}

function parseActionField(field, lineNumber) {
  var kv = parseKeyValue(field, lineNumber);
  if (kv.key === "点击") {
    return kv.value.split(",").map(function (name) {
      name = name.trim();
      if (name === "AUTO") {
        return { type: models.ActionType.CLICK_AUTO };
      }
      if (name === "BOSS") {
        return { type: models.ActionType.BOSS };
      }
      return { type: models.ActionType.CLICK_ROLE, role: name };
    });
  }
  if (kv.key === "提示") {
    return [{ type: models.ActionType.NOTIFY, message: kv.value }];
  }
  if (kv.key === "AUTO") {
    return [{ type: models.ActionType.TOGGLE_AUTO, value: kv.value === "开" ? "on" : "off" }];
  }
  if (kv.key === "SET") {
    return [{ type: "setRoles", values: kv.value.split(",").map(function (item) { return item.trim(); }) }];
  }
  throw new Error("第" + lineNumber + "行未知字段：" + kv.key);
}

function parseAxisText(text) {
  var lines = String(text).replace(/\r\n/g, "\n").split("\n");
  var header = {};
  var events = [];
  var inAxis = false;
  for (var i = 0; i < lines.length; i += 1) {
    var lineNumber = i + 1;
    var line = lines[i].trim();
    if (line === "" || line.indexOf("#") === 0) {
      continue;
    }
    if (line === "[轴]") {
      inAxis = true;
      continue;
    }
    if (!inAxis) {
      var headerKv = parseKeyValue(line, lineNumber);
      header[headerKv.key] = headerKv.value;
      continue;
    }
    var parts = line.split("|").map(function (part) { return part.trim(); });
    var timeSeconds = parseTimeToSeconds(parts[0]);
    var actions = [];
    for (var p = 1; p < parts.length; p += 1) {
      actions = actions.concat(parseActionField(parts[p], lineNumber));
    }
    events.push({
      id: "line-" + lineNumber,
      sourceLine: lineNumber,
      timeSeconds: timeSeconds,
      actions: actions
    });
  }
  return {
    type: header["轴类型"] === "开关" ? models.AxisType.SWITCH : models.AxisType.SEQUENCE,
    clickIntervalMs: parseInt(header["点击间隔"] || "100", 10),
    header: header,
    events: events
  };
}

module.exports = {
  parseAxisText: parseAxisText,
  parseTimeToSeconds: parseTimeToSeconds
};

File: android/app/src/main/java/com/kokkoro/clanbattle/axis/AxisParser.kt
package com.kokkoro.clanbattle.axis

import com.kokkoro.clanbattle.recognition.CharacterRole

object AxisParser {
    fun parse(text: String): AxisDocument {
        val header = linkedMapOf<String, String>()
        val events = mutableListOf<AxisEvent>()
        val switchOpenings = mutableListOf<SwitchAxisOpening>()
        val switchNodes = mutableListOf<SwitchAxisNode>()
        var inAxis = false
        var inSwitchAxis = false

        text.replace("\r\n", "\n").lines().forEachIndexed { index, source ->
            val lineNumber = index + 1
            val line = source.trim()
            if (line.isEmpty() || line.startsWith('#')) return@forEachIndexed
            if (line == "[轴]") {
                inAxis = true
                inSwitchAxis = false
                return@forEachIndexed
            }

            if (line.startsWith("[轴开局]")) {
                val parts = line.split('|').map(String::trim)
                require(parts.first() == "[轴开局]") { "第${lineNumber}行轴开局格式错误" }
                switchOpenings += SwitchAxisOpening(
                    sourceLine = lineNumber,
                    target = parseSwitchTarget(parseFields(parts.drop(1), lineNumber))
                )
                inSwitchAxis = true
                inAxis = false
                return@forEachIndexed
            }

            if (!inAxis && !inSwitchAxis) {
                val (key, value) = parseKeyValue(line, lineNumber)
                header[key] = value
                return@forEachIndexed
            }

            if (inSwitchAxis) {
                val parts = line.split('|').map(String::trim)
                val fields = parseFields(parts.drop(1), lineNumber)
                switchNodes += SwitchAxisNode(
                    id = "switch-line-$lineNumber",
                    sourceLine = lineNumber,
                    timeSeconds = parseTime(parts.first()),
                    trigger = parseNodeTrigger(fields),
                    target = parseSwitchTarget(fields)
                )
                return@forEachIndexed
            }

            val parts = line.split('|').map(String::trim)
            val sequenceFields = parts.drop(1).map { parseKeyValue(it, lineNumber) }
            val triggerFields = linkedMapOf<String, String>()
            sequenceFields.filter { (key, _) -> key in triggerKeys }.forEach { (key, value) ->
                require(triggerFields.put(key, value) == null) { "第${lineNumber}行字段重复：$key" }
            }
            val actions = sequenceFields
                .filterNot { (key, _) -> key in triggerKeys }
                .flatMap { (key, value) -> parseAction("$key=$value", lineNumber) }
            events += AxisEvent(
                id = "line-$lineNumber",
                sourceLine = lineNumber,
                timeSeconds = parseTime(parts.first()),
                actions = actions,
                trigger = parseNodeTrigger(triggerFields)
            )
        }

        return AxisDocument(
            type = if (header["轴类型"] == "开关" || switchOpenings.isNotEmpty()) {
                AxisType.SWITCH
            } else {
                AxisType.SEQUENCE
            },
            clickIntervalMs = header["点击间隔"]?.toIntOrNull() ?: 100,
            header = header,
            events = events,
            hasAxisSection = inAxis || switchOpenings.isNotEmpty(),
            switchOpenings = switchOpenings,
            switchNodes = switchNodes,
            hasSwitchSection = switchOpenings.isNotEmpty()
        )
    }

    fun parseTime(raw: String): Int {
        val match = Regex("^(\\d):([0-5]\\d)$").matchEntire(raw.trim())
            ?: error("时间格式必须为 M:SS：$raw")
        return match.groupValues[1].toInt() * 60 + match.groupValues[2].toInt()
    }

    private fun parseAction(field: String, lineNumber: Int): List<AxisAction> {
        val (key, value) = parseKeyValue(field, lineNumber)
        return when (key) {
            "点击" -> value.split(',').map { name ->
                when (val trimmed = name.trim()) {
                    "AUTO" -> AxisAction(ActionType.CLICK_AUTO)
                    "BOSS" -> AxisAction(ActionType.BOSS)
                    else -> AxisAction(ActionType.CLICK_ROLE, role = trimmed)
                }
            }

            "提示" -> listOf(AxisAction(ActionType.NOTIFY, message = value))
            "AUTO" -> listOf(
                AxisAction(
                    ActionType.TOGGLE_AUTO,
                    value = when (value) {
                        "开" -> "on"
                        "关" -> "off"
                        else -> null
                    },
                    rawValue = value
                )
            )
            "SET" -> listOf(AxisAction(ActionType.SET_ROLES, values = value.split(',').map(String::trim)))
            else -> error("第${lineNumber}行未知字段：$key")
        }
    }

    private fun parseKeyValue(line: String, lineNumber: Int): Pair<String, String> {
        val index = line.indexOf('=')
        require(index >= 0) { "第${lineNumber}行缺少等号：$line" }
        return line.substring(0, index).trim() to line.substring(index + 1).trim()
    }

    private fun parseFields(parts: List<String>, lineNumber: Int): Map<String, String> {
        val fields = linkedMapOf<String, String>()
        parts.filter(String::isNotBlank).forEach { part ->
            val (key, value) = parseKeyValue(part, lineNumber)
            require(fields.put(key, value) == null) { "第${lineNumber}行字段重复：$key" }
        }
        return fields
    }

    private fun parseSwitchTarget(fields: Map<String, String>): SwitchControlTarget {
        val rawRoles = fields["SET"]?.split(',')?.map(String::trim).orEmpty()
        val roles = CharacterRole.entries.zip(rawRoles.map(::parseToggleState)).toMap()
        val rawAuto = fields["AUTO"]
        return SwitchControlTarget(
            auto = rawAuto?.let(::parseToggleState),
            roles = roles,
            rawAuto = rawAuto,
            rawRoles = rawRoles,
            message = fields["提示"]
        )
    }

    private fun parseNodeTrigger(fields: Map<String, String>): SwitchNodeTrigger {
        val ubAfter = fields["UB后"]
        val pauseFrame = fields["卡帧"]
        if (ubAfter != null && pauseFrame != null) {
            return ConflictingSwitchTrigger(ubAfter, pauseFrame)
        }
        if (pauseFrame != null) {
            return PauseFrameTrigger(parseRole(pauseFrame), pauseFrame)
        }
        if (ubAfter == null) return TimedTrigger
        if (ubAfter.equals("BOSS", ignoreCase = true)) {
            val rawDelay = fields["延迟"]
            return BossDelayTrigger(
                minimumDelayMs = rawDelay?.toDoubleOrNull()?.let { (it * 1_000).toLong() },
                rawDelay = rawDelay
            )
        }
        return CharacterUbTrigger(parseRole(ubAfter), ubAfter)
    }

    private fun parseToggleState(raw: String): AxisToggleState? = when (raw) {
        "开" -> AxisToggleState.ON
        "关" -> AxisToggleState.OFF
        else -> null
    }

    private fun parseRole(raw: String): CharacterRole? = when (raw) {
        "角色1" -> CharacterRole.ROLE_1
        "角色2" -> CharacterRole.ROLE_2
        "角色3" -> CharacterRole.ROLE_3
        "角色4" -> CharacterRole.ROLE_4
        "角色5" -> CharacterRole.ROLE_5
        else -> null
    }

    private val triggerKeys = setOf("UB后", "延迟", "卡帧")
}
